### ISSUE TITLE
Difference in genome size estimation by genomescope depending on ploidy.

### ISSUE THREAD

**Issue by HenrivdGeest** [KamilSJaron/smudgeplot#80]
I have troubles understanding my genomescope plots. 

We have an unknown plant, of which the genome size is estimated to be >1Gbase (public sources)
The plant is supposed to be a homozygous diploid.
I have Ilumina data, and with the kmer plot we are trying to assess the genome size.
However, depending on ploidy, it either estimates the genome to be ~400MB (ploidy=2) or ~800MB (ploidy=1)
I am wondering what is the truth and what can cause this difference.
ploidy=2:
```
GenomeScope version 2.0
input file = kmc_kmer21.hist
p = 2
k = 21
max_kmercov = 900
property                      min               max               
Homozygous (aa)               84.4346%          97.0617%          
Heterozygous (ab)             2.93833%          15.5654%          
Genome Haploid Length         406,244,680 bp    408,221,979 bp    
Genome Repeat Length          190,375,373 bp    191,301,980 bp    
Genome Unique Length          215,869,307 bp    216,920,000 bp    
Model Fit                     66.5013%          83.8051%          
Read Error Rate               0.181954%         0.181954%       
```
![image](https://user-images.githubusercontent.com/2330901/101622431-533b8300-3a17-11eb-811e-6dcace3e5252.png)

For ploidy=1 we get:
```
GenomeScope version 2.0
input file = kmc_kmer21.hist
p = 1
k = 21
max_kmercov = 900
property                      min               max               
Homozygous (a)                100%              100%              
Genome Haploid Length         812,316,935 bp    816,977,890 bp    
Genome Repeat Length          366,065,207 bp    368,165,635 bp    
Genome Unique Length          446,251,728 bp    448,812,255 bp    
Model Fit                     63.6145%          92.2841%          
Read Error Rate               0.180435%         0.180435%      
```
![image](https://user-images.githubusercontent.com/2330901/101622534-76fec900-3a17-11eb-9b22-a7f334b77e3c.png)

A few extra remarks:
- the model fit seems low(66-83% or 63-92%). What is acceptable?
- The kcov value is always at the first peak, for other samples, this is sometimes halve the value of the first peak?
- There seems to be a second peak, which might mean that most of my genome is heterozygous (first peak) and only a small part homozygous (second peak) which I find hard to believe. Maybe the second peak are repeats with twice the occurence of unique parts?
- error rate seems to be between Q20(1%) and Q30 (0.1%) which I think is normal for Illumina sequencing.

**Comment by KamilSJaron** [comment]
Hi Henri,

thanks for opening this issue :-)

I will decompose the problem into 2 separated issues: 1. getting the right model; 2. getting a reliable genome size estimate.

1. GenomeScope fits a set of binomials to kmer spectra. Heterozygous genomes with high coverage and low rate of sequencing errors are easy. However, sometimes when the genome is too homozygous / coverage is not great / error rate is high, genomescope misses the first peak.

That's exactly what must have happened in your case. Knowing it's a homozygous diploid is actually really useful. This tells us about 2 things: 1. the first (heterozygosity) peak must be rather small 2. the major peak in the kmer spectra will be the 2n peak. However, in your diploid fit it seems like 1n was estimated to be 46.8x, which looks wrong. The peak is by far too high to be the heterozygosity peak in your genome. It should have been estimated to ~23.4x.

You can give GenomeScope a prior for coverage, when I run it as `genomescope.R -i c8.hist -o . -n plant -k 31 -l 26`, I got the model accurately fitting individual genomic peaks.

![plant_plot](https://user-images.githubusercontent.com/8181573/101677343-5bfe7a00-3a54-11eb-8716-e38cf6fb9ace.png)

The model is fairly similar to the haploid model you made (concerning genome size), but unlike the haploid model, it actually reflects the diploid nature of the genome.

2. Getting the right genome size. I noticed you stopped counting kmers after 1000x. Maybe it won't matter in your case, but for genomes >500M it usually is a problem. [Look at this extreme case of marbled crayfish genome](https://kamilsjaron.github.io/peculiar-genomic-observations/biological/2020/01/crayfish.html) that has so super repetitive genome that ignoring kmers with higher than 10000x coverage caused that we missed about a half of the genome. My guess would be that if you take those kmers into account, you will get a >1G genome size estimate.

Just tiny note, the original idea behind ignoring high frequency kmers is to exclude chloroplats, mt DNA and contamination. But in my experience, those affect the genome size estimates much less than ignoring real genomic repetitions.

**Specific questions**

As the model fits (66-83% or 63-92%). Honestly, I am not sure. I don't really like genomescope confidence intervals and I don't look at the much, because they are missing what I think is the main problem of genomescope - the assumption of random distribution of heterozygous nucleotides across the genome. My gut feeling is telling me that will affect the estimates much more than unexplained variability of the model

> The kcov value is always at the first peak, for other samples, this is sometimes halve the value of the first peak?

Yes, kcov is 1n coverage estimate. If it's half of the "first peak", it means that the big one is not really the first one (i.e. there are only too few heterozygous kmers to form a proper detectable peak, but it should be there).

> error rate seems to be between Q20(1%) and Q30 (0.1%) which I think is normal for Illumina sequencing.

This is not the best, not the worst library I have ever seen. It looks to me like 7 rounds of PCR during library prep? The error is not just about the absolute frequency of errors but also about their coverage variations. PCR steps generate amplified sequencing errors that make those error tails.

Does it all make sense?

**Comment by mschatz** [comment]
Thanks Kamil! I couldnt have said it better myself!

Mike

**Comment by HenrivdGeest** [comment]
Thanks for the explanations, I do get it much better now. For now I will make sure the kcov values is at the first 'peak' I clearly see and where I expect the peak to be.  For several pacbio hifi data-sets I see the same thing happening, but with setting the proper kcov, its first the genome size estimation much better. I know max kmercov was made for plastid 'contamination' filtering, but for my plants it does not seem to matter that much where I put the cutoff. (around 10% in genome size difference).
I am not sure on the illumina data, how that is made, but it could be a pcr based library prep.

**Comment by mschatz** [comment]
Great. Yes, it is pretty common to have a ~10% variability in genome size
depending on where you draw the cutoff to account for plastid sequencing,
phiX, and other "contamination" in the data. Depending on the tissue used
for sequencing, we often see some levels of bacterial contamination as well
-- for our big tomato genome project earlier this year we detected a few
percent of bacterial reads in 2 of the 14 samples we assembled, for
example. The details are described in the "MAS2.0 Genome Assembly" section
of the Star Methods here (page e6):
https://www.cell.com/cell/fulltext/S0092-8674(20)30616-4

Mike
